dipy/direction/closest_peak.py

```diff
@@ -0,0 +1,213 @@
+        self._B, m, n = basis(sh_order, sphere.theta, sphere.phi)
+
+    def get_pmf(self, point):
+        coeff = trilinear_interpolate4d(self.shcoeff, point)
+        pmf = np.dot(self._B, coeff)
+        return pmf
+
+
+class BaseDirectionGetter(DirectionGetter):
+    """A base class for dynamic direction getters"""
+
+    def __init__(self, pmf_gen, max_angle, sphere, pmf_threshold=.1, **kwargs):
+        self.sphere = sphere
+        self._pf_kwargs = kwargs
+        self.pmf_gen = pmf_gen
+        if pmf_threshold < 0:
+            raise ValueError("pmf threshould must be >= 0.")
+        self.pmf_threshold = pmf_threshold
+        self.cos_similarity = np.cos(np.deg2rad(max_angle))
+
+    def _peak_directions(self, blob):
+        """Gets directions using parameters provided at init.
+
+        Blob can be any function defined on ``self.sphere``, ie an ODF, PMF,
+        FOD.
+        """
+        return peak_directions(blob, self.sphere, **self._pf_kwargs)[0]
+
+    def initial_direction(self, point):
+        """Returns best directions at seed location to start tracking.
+
+        Parameters
+        ----------
+        point : ndarray, shape (3,)
+            The point in an image at which to lookup tracking directions.
+
+        Returns
+        -------
+        directions : ndarray, shape (N, 3)
+            Possible tracking directions from point. ``N`` may be 0, all
+            directions should be unique.
+
+        """
+        pmf = self.pmf_gen.get_pmf(point)
+        pmf.clip(min=self.pmf_threshold, out=pmf)
+        return self._peak_directions(pmf)
+
+    def get_direction(self, point, direction):
+        pmf = self.pmf_gen.get_pmf(point)
+        pmf.clip(min=self.pmf_threshold, out=pmf) 
+        peaks = self._peak_directions(pmf)
+        if len(peaks) == 0:
+            return 1
+        new_dir = _closest_peak(peaks, direction, self.cos_similarity)
+        if new_dir is None:
+            return 1
+        else:
+            direction[:] = new_dir
+            return 0
+
+
+class ClosestPeakDirectionGetter(BaseDirectionGetter):
+    """A direction getter that returns the closest odf peak to previous tracking
+    direction.
+
+    """
+    @classmethod
+    def from_pmf(klass, pmf, max_angle, sphere=default_sphere, pmf_threshold=0.1,
+                 **kwargs):
+        """Constructor for making a DirectionGetter from an array of Pmfs
+
+        Parameters
+        ----------
+        pmf : array, 4d
+            The pmf to be used for tracking at each voxel.
+        max_angle : float, [0, 90]
+            The maximum allowed angle between incoming direction and new
+            direction.
+        sphere : Sphere
+            The set of directions to be used for tracking.
+        pmf_threshold : float [0., 1.]
+            Used to remove direction from the probability mass function for
+            selecting the tracking direction.
+        relative_peak_threshold : float in [0., 1.]
```

Comment:
unexpected parameter in docstring. It should be removed or added as function parameter

Reply:
Consumed par parent constructor.